### ISSUE TITLE
Refactor `fs_iso9660`

### ISSUE THREAD

**Issue by QuzarDC** [KallistiOS/KallistiOS#767]
There would be two interrelated changes I'm imagining here:

- Restructure `fs_iso9660` to use the mounting paradigm like fat/ext2 do. By default create the mount to `/cd/` as we've always done, but this would allow for more flexible mounting of `fs_iso9660`.

- Use `kos_blockdev_t` as an intermediary between the cdrom driver and fs. This may require enhancement to the blockdev system, but would facilitate both adding new FSes that could be read from the cdrom (maybe HFS, UDF, VCD, or proprietary formats) and easily mounting disc images (from sd, g1ata, or an existing mounted fs). It could also facilitate the writing of an SPI/ATAPI based cdrom driver (or drivers for alternate hardware like a g1ata CDROM).

I wanted to be sure to float the idea for discussion and feedback before taking any action, as well as offering it up publicly in case anyone else has interest in doing this (or it would aid any existing uses).

**Comment by DC-SWAT** [comment]
This is a great idea. Plus I have a working driver for optical drives in the ISO Loader firmware.
But it is better not to replace the work with the standard GD drive, only for other slave drives like DVD for example.

**Comment by Tchan0** [comment]
Indirectly (?) related to that, it would be nice to be able to use the iso9660 file system when only a virtual cdrom (over network/serial) is present, but no hardware cd/gdrom reader is present (eg: Naomi without gd-rom, System SP)

**Comment by QuzarDC** [comment]
> Indirectly (?) related to that, it would be nice to be able to use the iso9660 file system when only a virtual cdrom (over network/serial) is present, but no hardware cd/gdrom reader is present (eg: Naomi without gd-rom, System SP)

That should be possible with this by providing those virtual cdroms as a different backing blockdev, in the same way that you'd mount an iso9660 from a file, for instance. :)

**Comment by DC-SWAT** [comment]
> > Indirectly (?) related to that, it would be nice to be able to use the iso9660 file system when only a virtual cdrom (over network/serial) is present, but no hardware cd/gdrom reader is present (eg: Naomi without gd-rom, System SP)
> 
> That should be possible with this by providing those virtual cdroms as a different backing blockdev, in the same way that you'd mount an iso9660 from a file, for instance. :)

Yes, but in that case, what I'm working on in #763 requires some work for blockdev too, apparently.
Because the current blockdev provides access only for reading sectors. This also applies to IDE by the way, streams is also can be implemented there as for the GD drive.

**Comment by pcercuei** [comment]
One update that's greatly needed for the `fs_iso9660` code is caching of the TOC. Having it uncached means the lens has to move all the way to the TOC area then back to the file's location on the disc every time a file is opened.

**Comment by DC-SWAT** [comment]
> One update that's greatly needed for the `fs_iso9660` code is caching of the TOC. Having it uncached means the lens has to move all the way to the TOC area then back to the file's location on the disc every time a file is opened.

It is cached now, 16 sectors for inode data. It's just that when there are a lot of files, this cache is not enough. The cache needs to be made more compact and voluminous.
